### PR TITLE
Added product image upload method in products.php api file.

### PR DESCRIPTION
**Pull request by surajmandal04**
Added method to upload image in product give by product id.

### CODE REVIEW

File: src/ShoppinPal/Vend/Api/V2/Products.php [hunk 1]
```diff
@@ -68,4 +68,33 @@ public function get($productId)
...
+     * Returns the true, that update image.
+     *
+     * @param string $productId ID of the product.
+     * @param string $imagePath of the image.
+     *
+     * @return true
+     *
+     * @throws EntityNotFoundException If the product is not found.
+     */
+    public function imageUpload($productId, $imagePath)
```
Comment (szeber) [inline src/ShoppinPal/Vend/Api/V2/Products.php]: You should expect a file stream resource or the content of the image here. The image may not be on the filesystem.

File: src/ShoppinPal/Vend/Api/V2/Products.php [hunk 2]
```diff
@@ -68,4 +68,33 @@ public function get($productId)
...
+    {
+        $request = $this->getAuthenticatedRequestForUri('api/2.0/products/' . urlencode($productId) . '/actions/image_upload');
+
+        $request->setMethod(CurlHttpRequest::METHOD_POST);
+
+        $request->addHeader('Content-Type: multipart/form-data');
+
+        $request->setPayload(
+            json_encode(array("image"=>fopen($imagePath, 'r'))),
```
Comment (szeber) [inline src/ShoppinPal/Vend/Api/V2/Products.php]: You need to send a multipart/form data request, that doesn't expect a json. So json_encoding makes no sense.

Comment (szeber) [reply src/ShoppinPal/Vend/Api/V2/Products.php]: fopen creates a file handle resource. You can't send that to Vend. You need the actual image data.

Comment (szeber) [reply src/ShoppinPal/Vend/Api/V2/Products.php]: You need to create a valid multipart/form request string here if you use raw. Raw means it send the provided string as the request body without any modifications.

**Review by szeber — CHANGES_REQUESTED** [review]

File: src/ShoppinPal/Vend/Api/V2/Products.php (hunk 1 above)
Comment (emulgeator) [reply src/ShoppinPal/Vend/Api/V2/Products.php]: I'd rather call thus uploadImage to follow the convention.

File: src/ShoppinPal/Vend/Api/V2/Products.php [hunk 3]
```diff
@@ -68,4 +68,33 @@ public function get($productId)
         return new Product($result['data'], Product::UNKNOWN_PROPERTY_IGNORE, true);
 
     }
+
+    /**
+     * Returns the true, that update image.
```
Comment (emulgeator) [inline src/ShoppinPal/Vend/Api/V2/Products.php]: This doc does not make much sense

File: src/ShoppinPal/Vend/Api/V2/Products.php [hunk 4]
```diff
@@ -68,4 +68,33 @@ public function get($productId)
...
+        $request->addHeader('Content-Type: multipart/form-data');
+
+        $request->setPayload(
+            json_encode(array("image"=>fopen($imagePath, 'r'))),
+            CurlHttpRequest::PAYLOAD_TYPE_RAW
+        );
+
+        $result = $this->sendRequest($request, 'product image upload');
+
+        return true;
```
Comment (emulgeator) [inline src/ShoppinPal/Vend/Api/V2/Products.php]: If the returned value makes no sense don't return anything. just make a void method

File: src/ShoppinPal/Vend/Api/V2/Products.php (hunk 3 above)
Comment (emulgeator) [reply src/ShoppinPal/Vend/Api/V2/Products.php]: Doc still need to be changed. Something like "Uploads the given image of the given product" would be more appropriate.

File: src/ShoppinPal/Vend/Api/V2/Products.php [hunk 5]
```diff
@@ -68,4 +68,37 @@ public function get($productId)
...
+    {
+        $request = $this->getAuthenticatedRequestForUri('api/2.0/products/' . urlencode($productId) . '/actions/image_upload', [], true);
+
+        $request->setMethod(CurlHttpRequest::METHOD_POST);
+
+        $boundary = uniqid();
+
+        $request->addHeader("Content-Type: multipart/form-data; boundary=----".$boundary);
+
+        $body = "------".$boundary."\r\n"."Content-Disposition: form-data; name=\"image\"; filename=\"upload-image.jpg\"\r\nContent-Type: application/octet-stream\r\n\r\n".$imageData."\r\n\r\n\r\n------".$boundary."--";
```
Comment (emulgeator) [inline src/ShoppinPal/Vend/Api/V2/Products.php]: This string is already concatenated, please wrap it into multiple rows by block to make it more readable.
Also you use " for a string which contains "-s therefore you need to escape them. Use ' instead what's the default anyway.

`
        $request = $this->getAuthenticatedRequestForUri('api/2.0/products/' . urlencode($productId) . '/actions/image_upload', [], true);
        $request->setMethod(CurlHttpRequest::METHOD_POST);
        $boundary = uniqid();
        $request->addHeader('Content-Type: multipart/form-data; boundary=----' . $boundary);

        $eol = "\r\n";

        $body =
            '------' . $boundary . $eol
            . 'Content-Disposition: form-data; name="image"; filename"upload-image.jpg"' . $eol
            . 'Content-Type: application/octet-stream' . $eol . $eol
            . $imageData . $eol . $eol . $eol
            . '------' . $boundary . '--';
`

Comment (emulgeator) [reply src/ShoppinPal/Vend/Api/V2/Products.php]: Also concatenation is a simple operator just like + or -, spaces should be added around the operators

File: src/ShoppinPal/Vend/DataObject/Entity/V2/ProductImageUpload.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace ShoppinPal\Vend\DataObject\Entity\V2;
+
+use ShoppinPal\Vend\DataObject\Entity\EntityDoAbstract;
+
+class ProductImageUpload extends EntityDoAbstract
+{
+    public $id;
```
Comment (emulgeator) [inline src/ShoppinPal/Vend/DataObject/Entity/V2/ProductImageUpload.php]: Document the parameters, at least their type

**Review by emulgeator — CHANGES_REQUESTED** [review]

File: src/ShoppinPal/Vend/Api/V2/Products.php (hunk 4 above)
Comment (surajmandal04) [reply src/ShoppinPal/Vend/Api/V2/Products.php]: In storage service need the response data. So return the response data.

File: src/ShoppinPal/Vend/Api/V2/Products.php [hunk 7]
```diff
@@ -68,4 +68,43 @@ public function get($productId)
...
 
     }
+
+    /**
+     * Uploads the give image of given product and
+     * Return response of product image upload.
+     * @param string $productId ID of the product.
+     * @param string $imageData of the image.
+     *
+     * @return data
```
Comment (emulgeator) [inline src/ShoppinPal/Vend/Api/V2/Products.php]: data type does not exist. @return should contain the exact type of the returned value in this case ProductImageUpload

File: src/ShoppinPal/Vend/DataObject/Entity/V2/ProductImageUpload.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace ShoppinPal\Vend\DataObject\Entity\V2;
+
+use ShoppinPal\Vend\DataObject\Entity\EntityDoAbstract;
+
+class ProductImageUpload extends EntityDoAbstract
+{
+	/** @var string */
```
Comment (emulgeator) [inline src/ShoppinPal/Vend/DataObject/Entity/V2/ProductImageUpload.php]: Indentation issue

**Review by emulgeator — CHANGES_REQUESTED** [review]
A few small changes still needed

File: src/ShoppinPal/Vend/DataObject/Entity/V2/ProductImageUpload.php (hunk 8 above)
Comment (surajmandal04) [reply src/ShoppinPal/Vend/DataObject/Entity/V2/ProductImageUpload.php]: Resolved

File: src/ShoppinPal/Vend/Api/V2/Products.php (hunk 7 above)
Comment (surajmandal04) [reply src/ShoppinPal/Vend/Api/V2/Products.php]: Added return type

**Review by emulgeator — APPROVED** [review]

**Pull request merged**